PROLL-172: New player UI

TheMankel opened the pull request:
### This PR adds a new design to the midi player UI


---
[Task link](https://sacrebleu.atlassian.net/browse/PROLL-172)

roszcz commented:
Thanks, this looks great! 🚀 

I just added another example dashboard so you can test with more realistic examples. It requires a new python dependency:

```sh
pip install datasets

streamlit run streamlit_pianoroll/example_maestro.py
```

You should get something like this:

<img width="420" alt="image" src="https://github.com/Nospoko/streamlit-pianoroll/assets/8056825/56e7a413-4406-42b8-8353-5da7a10c42be">

In terms of the UI, here's an idea to consider:

1. Show a YouTube-like progress indicator (I think this PR started with something like that) that is "connected" to the lower pianoroll, and moving one moves the other.
2. Make the visibility of the lower piano roll optional, i.e. add an argument to the python interface, so it can be called like:

```python
from_fortepyan(piece, show_bird_view=False)
```
and don't show the lower pianoroll, so any action can be done only on the "line and dot" seekbar.

The motivation for this is that for MIDI research purposes it will often be the case that the MIDI sequence will be one "page" long, and in that case the lower pianoroll is just a duplicate of the upper.

Let me know if this makes sense :)

TheMankel commented:
This idea makes sense. I have added such an option.

https://github.com/Nospoko/streamlit-pianoroll/assets/26410548/4d72b3fb-152d-430d-85ce-6008b6d8183a

I found a strange bug - when we move only the default seekbar, the track changes currentTime correctly. But when we stop playing after some time, and then start playing again, when we move the default seekbar, it adds an offset to the timeline. The offset value corresponds to the time we have already played. Here is a video with such bug

https://github.com/Nospoko/streamlit-pianoroll/assets/26410548/551918cf-2895-42e5-a381-1339239b6c79

I've tested this behavior on the pianoroll project as well and it's the same, so I'm guessing it's something wrong with the library itself and not just after my changes 🙈 From what I've found, when we don't move the default seekbar, but only use the added lower pianoroll timeline, the bug doesn't occur. The difference in the code is that I use the `stop()` method to control the progress timeline, while the library uses the `pause()` method to control the seekbar. When I changed `stop()` to `pause()` the bug occurred again. So I will try to find a solution to this problem, but it seems very tricky to me

roszcz commented:
>  The difference in the code is that I use the stop() method to control the progress timeline, while the library uses the pause() method to control the seekbar. When I changed stop() to pause() the bug occurred again. So I will try to find a solution to this problem, but it seems very tricky to me

Is it possible to implement the seek bar outside of that library and make sure it uses `stop()` that way? :)

TheMankel commented:
I added a custom seekbar and I think it also solved the problem I wrote about earlier 

> The problem I noticed is the player doesn't always stop playing when it gets to the end of the track

https://github.com/Nospoko/streamlit-pianoroll/assets/26410548/34994d58-a376-4d37-b3de-a2a98bd6d480

roszcz commented:
Thanks, looks awesome 😍 

I noticed that when using the seek bar (either one) the play button shows briefly before resuming playback (it happens in your video) - is this something we can control?

TheMankel commented:
I added these two event listeners to display and hide the central play button. I can restore the earlier version where this button did not appear when using the seekbar. I will have to check if the button appears when the recording reaches the end and adjust the code if necessary.

```js
this.midiPlayer.addEventListener("stop", () => {
  centralPlayButton.classList.remove("fadeOut")
  centralPlayButton.classList.add("fadeIn")
})

this.midiPlayer.addEventListener("start", () => {
  centralPlayButton.classList.remove("fadeIn")
  centralPlayButton.classList.add("fadeOut")

  this.isPlaying = true
})
```

roszcz commented on streamlit_pianoroll/frontend/public/styles.css:
```diff
@@ -3,15 +3,24 @@
+.pianoroll-progress-bar {
+  background-color: whitesmoke;
+  width: 100%;
+  height: 6.25rem;
```
Can we make this relative to the upper svg height? In the standard streamlit view it looks great for me with `height: 3.00rem;`:

<img width="420" alt="image" src="https://github.com/Nospoko/streamlit-pianoroll/assets/8056825/9464698e-f88e-404b-b7d1-5da07d2b2ff6">

But in the "wide mode" I think 5rem looks better and is similar in ratio:

<img width="420" alt="image" src="https://github.com/Nospoko/streamlit-pianoroll/assets/8056825/165eac9e-5b09-4c9a-bcfa-4a6d3abebfb8">

TheMankel commented:
There is a problem with implementing this idea. It seems to me that streamlit does not offer a way to check whether "wide mode" is activated. I was only able to come up with a solution on the frontend side, which works only if the application is built. In the case of localhost, such a solution will not work, because we do not meet the principle of same-origin-policy due to different ports.

In `main.ts`:
```js
function detectWideMode() {
  try {
    const progressBarSVG = document.querySelector("#progress-bar") as SVGElement
    if (progressBarSVG) {
      console.log(window.parent.origin)
      console.log(window.parent.document.querySelector("div[data-layout]"))
      const element = window.parent.document.querySelector(
        "div[data-layout]"
      ) as HTMLDivElement
      console.log(element.getAttribute("data-layout"))
      if (element.getAttribute("data-layout") === "narrow") {
        console.log("narrow")
        progressBarSVG.setAttribute("height", "3rem")
      } else if (element.getAttribute("data-layout") === "wide") {
        console.log("wide")
        progressBarSVG.setAttribute("height", "5rem")
      }
    }
  } catch (error) {
    console.log(error)
  }
}
```

Then we need to call this method at the beginning of `export function onStreamlitRender(event: Event): Void`, and also inside the first condition in the `if (showBirdView)` statement.

I'm not sure if this satisfies you, as on localhost it doesn't work

roszcz commented:
I see, is there no way to set the height in terms of the percentage of the upper pianoroll, or the parent element?

TheMankel commented:
Actually, it is possible. I wrote a function for this:
```js
function updateProgressSvgHeight() {
  const pianoRollSvgVisualizer = document.querySelector(
    "#my-svg"
  )! as MidiPlayerElement
  const progressBarSVG = document.querySelector(
    "#progress-bar"
  )! as MidiPlayerElement

  if (!pianoRollSvgVisualizer || !progressBarSVG) return

  const percentage = 16.5
  const minHeight = 48 // 3rem if font-size is 16px
  const maxHeight = 80 // 5rem if font-size is 16px

  let progressHeight =
    (percentage / 100) * pianoRollSvgVisualizer.getBoundingClientRect().height

  progressHeight = Math.max(minHeight, Math.min(progressHeight, maxHeight))
  console.log(progressHeight.toFixed(0))
  progressBarSVG.setAttribute("height", `${progressHeight.toFixed(0)}px`)
}
```
So by changing `percentage`, `minHeight` and `maxHeight` you can adjust it to your needs



https://github.com/Nospoko/streamlit-pianoroll/assets/26410548/291c8182-d68c-47ab-bc68-e64c118e62e6

roszcz commented on streamlit_pianoroll/frontend/src/main.ts:
```diff
@@ -122,3 +158,23 @@ export function onStreamlitRender(event: Event): void {
+
+  const percentage = 16.5
+  const minHeight = 48 // 3rem if font-size is 16px
+  const maxHeight = 80 // 5rem if font-size is 16px
```
```suggestion
  const maxHeight = 220 // 13.75rem if font-size is 16px
```

I think it makes sense for full screen mode to allow more range here :)
